invenio_circulation/transitions/base.py

```diff
@@ -8,59 +8,111 @@
+        missing = [p for p in self.REQUIRED_PARAMS if p not in kwargs]
+        if missing:
+            msg = 'Required input parameters are missing `[{}]`'\
+                .format(missing)
+            raise TransitionConstraintsViolation(msg=msg)
+        return f(self, loan, **kwargs)
+    return inner
 
 
 def check_trigger(f):
-    """."""
+    """Decorator to check the transition should be manually triggered."""
     def inner(self, loan, **kwargs):
-        """."""
         if kwargs.get('trigger', 'next') != self.trigger:
-            raise TransitionValidationFailed(
-                msg='No param `trigger` with value `{0}` found.'
-                    .format(self.trigger))
+            msg = 'No param `trigger` with value `{0}`.'.format(self.trigger)
+            raise TransitionConditionsFailed(msg=msg)
         return f(self, loan, **kwargs)
     return inner
 
 
 class Transition(object):
     """A transition object that is triggered on conditions."""
 
+    REQUIRED_PARAMS = [
+        'transaction_user_pid',
+        'patron_pid',
+        'transaction_location_pid',
+        'transaction_date',
+        'item_pid'
+    ]
+
     def __init__(self, src, dest, trigger='next', permission_factory=None,
                  **kwargs):
-        """."""
+        """Init transition object."""
         self.src = src
         self.dest = dest
         self.trigger = trigger
         self.permission_factory = permission_factory
+        # validate states
+        self.validate_transition_states()
+
+    def validate_transition_states(self):
+        """Ensure that source and destination states are valid."""
+        states = current_app.config['CIRCULATION_LOAN_TRANSITIONS'].keys()
+        if not all([self.src in states, self.dest in states]):
+            msg = 'Source state `{0}` or destination state `{1}` not in [{2}]'\
+                .format(self.src, self.dest, states)
+            raise InvalidState(msg=msg)
 
+    @ensure_same_item_patron
+    @ensure_required_params
     @check_trigger
     def before(self, loan, **kwargs):
-        """Evaluate conditions and raise if anything wrong."""
-        if not all([self.src in current_circulation.circulation.transitions,
-                    self.dest in current_circulation.circulation.transitions]):
-            raise TransitionValidationFailed(
-                msg='Source state `{0}` or destination state `{1}` not in '
-                    '[{2}]'.format(
-                        self.src,
-                        self.dest,
-                        current_circulation.circulation.transitions.keys()
-                    )
-            )
-
-    def update_loan(self, loan, **kwargs):
-        """Update current loan with new values."""
+        """Validate input, evaluate conditions and raise if failed."""
+        kwargs.setdefault('transaction_date', datetime.now())
+        kwargs['transaction_date'] = parse_date(kwargs['transaction_date'])
         loan.update(kwargs)
```

Comment:
I had a long thought about that, the issue is that you cannot update/modify kwargs in a function and expect they are updated in next execution, because kwargs are copied, new dict.
So, for example, you cannot easily add/set the start and end dates in the kwargs if not passed in the request.
Then I was think if we should not update the loan and I did not find any reason not to do that, even if at the beginning I was happy to keep loan and kwargs separated in case we need to compare previous and new values.
I think that if we don't want to do that, u should change the code functions to have method(loan, event, **kwargs), where event is an object/dict with all the params passed through the request, and the u can easily modify it.
I am sorry but I don't have easy access to code while on vacation, I have to change it, feel free to adapt the code to what u think is best!!!

Reply:
@jma but there is no implication on the loan. It is just a dict update, no commit happened. IMHO I think is safe.

Reply:
@zzacharo ok